---
title: "230104_panda_network_inputs"
author: "Jennifer Fisher"
date: "2023-01-04"
output:
  html_document:
    toc: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# START 
    Goal/Purpose of operations: 
    For the gene regulatory analysis, I need to data wrangle the protein protein interaction data from STRING and the motif infromation that was created by Kimberly Class lab (https://sites.google.com/a/channing.harvard.edu/kimberlyglass/tools/resources?pli=1). 
    
    Finished psedocode on:
    230104
    
    System which operations were done on: 
    my laptop(on cheaha)
    
    GitHub Repo:
    Sex_Bias_Adverse_Events
    
    Directory of operations:
    /data/project/lasseigne_lab/JLF_scratch/Sex_Bias_Adverse_Events
    
    Scripts being edited for operations:
    NA
    
    Data being used:
    STRING database-  version 11.5- This database contains known and predicted protein-protein interactions, including direct and indirect associations 31. We divided the protein-protein interactions in the network by 1000 and filtered if the interaction was greater than 0.7. A score greater than 0.7 indicates high confidence in that protein-protein interaction. We converted the protein Ensembl ID was to the HGNC gene symbol for each protein using the provided STRING annotations.
    
    TF motif data-The mappings below were produced by using FIMO to scan a genome for sets of  CIS-BP Single Species DNA-motifs curated in the MEME suite. Hits that met a significance less than 1e-4 were recorded and mapped to genes based on annotations downloaded from UCSC. Significant motif hits that fell within [-750,+250] of a gene's transcriptional start site were used to construct a prior regulatory network.  https://sites.google.com/a/channing.harvard.edu/kimberlyglass/tools/resources?pli=1
    
    Papers and tools: 
    NA

# STEPS

### Set working directory 
```{r}
#change this path to match your paths 
dir_path <- "/data/project/lasseigne_lab/JLF_scratch/Sex_Bias_Adverse_Events/"
```

### load in data
Done within the script
### Analysis

```{r eval= FALSE}
download.file('https://stringdb-static.org/download/protein.links.v11.5/9606.protein.links.v11.5.txt.gz', destfile = paste0(dir_path, "data/panda_input_data/stringi_database_11_5_9606.txt.gz"), method = "wget", extra= "--no-check-certificate")
```
make sure to unzip the file outside of the script
```{r}
ppi_table<- read.delim(paste0(dir_path, "data/panda_input_data/stringi_database_11_5_9606.txt"), sep= " ", header = TRUE)
```

```{r}
head(ppi_table)
```

```{r}
hist(ppi_table$combined_score)
```
need to change to symbols and divide scores and self = 1

```{r}
ppi_table$combined_score<- ppi_table$combined_score /1000
```


```{r}
ppi_table$combined_score<- ifelse(ppi_table$protein1 == ppi_table$protein2, 1, ppi_table$combined_score)
```

```{r}
hist(ppi_table$combined_score)
```

```{r eval= FALSE}
download.file('https://stringdb-static.org/download/protein.info.v11.5/9606.protein.info.v11.5.txt.gz', destfile = paste0(dir_path, "data/panda_input_data/9606.protein.info.v11.5.txt.gz"), method = "wget", extra= "--no-check-certificate")
```

make sure to unzip file outside the script

```{r}
ppi_protein_info<- t(as.data.frame( strsplit(readLines("/data/project/lasseigne_lab/JLF_scratch/Sex_Bias_Adverse_Events/data/panda_input_data/9606.protein.info.v11.5.txt"), "\t")))
ppi_protein_info <- as.data.frame(ppi_protein_info)
```

```{r}
ppi<- ppi_table[ppi_table$combined_score >0.7,]
```


```{r}
protein1_symbols <- c()
protein2_symbols<- c()
for (i in 1:length(ppi$protein1)){
  #print(i)
  protein1_symbols[i]<- ppi_protein_info$V2[ppi$protein1[i] == ppi_protein_info$V1]
  protein2_symbols[i]<-ppi_protein_info$V2[ppi$protein2[i] == ppi_protein_info$V1]
}

```

```{r}
scores<- ppi$combined_score
symbol_ppi_table <- cbind(protein1_symbols, protein2_symbols, scores)
```

```{r}
symbol_ppi_table<- as.data.frame(symbol_ppi_table)
symbol_ppi_table$scores <- as.numeric(symbol_ppi_table$scores)
```


```{r eval =FALSE}
saveRDS(symbol_ppi_table, "/data/project/lasseigne_lab/JLF_scratch/Sex_Bias_Adverse_Events/data/panda_input_data/230104_symbol_ppi_network_11_5_0.rds")
```

```{r}
protein1_symbols[protein1_symbols == "NA"]
protein1_symbols[is.na(protein1_symbols)]
protein2_symbols[protein2_symbols == "NA"]
protein2_symbols[is.na(protein2_symbols)]
```

```{r}
head(symbol_ppi_table)
```

good

```{r}
motif_info_table<- read.table(paste0(dir_path,"data/panda_input_data/Homo_sapiens_motifinfo.txt"), sep= "\t", header = TRUE)
```
```{r}
head(motif_info_table)
```


```{r}
motif_table<- read.table(paste0(dir_path,"data/panda_input_data/hg38_refseq_-750_250_1e-4"), sep= "\t")
```

```{r}
head(motif_table)
```

```{r}
motif_info_table$MotifID
motif_info_table$id_trim <- t(as.data.frame(strsplit(motif_info_table$MotifID, "_", fixed = FALSE)))[,1]
```

```{r}
#motif_info_table$id_trim
motif_gene<- c()
for (i in 1:length(motif_table$V1)){
  motif_gene[i]<- motif_info_table$TF_Info[motif_table$V1[i] == motif_info_table$id_trim]
}
print("done")
```

```{r}
#motif_gene
motif_gene_v2 <- unlist(lapply(strsplit(motif_gene, "_", fixed = FALSE), `[[`, 1))
library(stringr)
motif_gene_v2<- str_replace(motif_gene_v2, "\\(", "")
motif_gene_v2<- str_replace(motif_gene_v2, "\\)", "")
```



```{r}
symbol_motif_table<- cbind(motif_gene_v2, motif_table$V2, motif_table$V3)
```

```{r}
head(symbol_motif_table)
```

```{r}
symbol_motif_table<- as.data.frame(symbol_motif_table)
symbol_motif_table$V3 <- as.numeric(symbol_motif_table$V3)
```

```{r eval =FALSE}
saveRDS(symbol_motif_table, paste0(dir_path, "data/panda_input_data/230104_symbol_motif_table.rds"))
```
### Save Data
Done in script
    
# END
    Location of final scripts:
    scripts
  
    
    Location of data produced:
    /data/project/lasseigne_lab/JLF_scratch/Sex_Bias_Adverse_Events/data/panda_input_data
    
    Dates when operations were done:
    230104
    
## Versions
```{r}
sessionInfo()
```


