---
title: "230313_sbae_drug_target"
author: "Jennifer Fisher"
date: "2023-03-13"
output:
  html_document:
    toc: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# START 
    Goal/Purpose of operations: 
    I am going to be looking at the enriched drug targets from the SBAE drugs and looking at gene expression and core genes of sex-specific communities. 
    
    Finished psedocode on:
    230308
    
    System which operations were done on: 
    my laptop
    
    GitHub Repo:
    230321_JLF_Sex_bias_adverse_events
    
    Docker: 
    jenfisher7/rstudio_sex_bias_drugs
    
    Directory of operations:
    ~/Documents/230321_JLF_Sex_bias_adverse_events
    
    Scripts being edited for operations:
    NA
    
    Data being used:
    output/alpaca/alpaca_runs
    
    Papers and tools: 
    "~/results/FARES_plots/drug_info_sbae_drugs.rds"- drug info from drugbank for SBAE drugs
    "~/results/FARES_plots/drug_info_all.rds"- drug info from drugbank for all FAERS drugs
    "~/results/FARES_plots/sbae_targets_df_top.rds"- the SBAE drug targets
    ~/results/alpaca/alpaca_core_gene_lists- the core gene sets

# STEPS

### Set working directory 
```{r}
source("~/src/sex_bias_drug_functions.R")
library(ggplot2)
library(circlize)
library(ComplexHeatmap)
library(drugbankR)
library(DBI)
library(hash)
library(org.Hs.eg.db)
library(EnsDb.Hsapiens.v75)
library(stringr)
```


```{r}
drug_info_sbae_drugs <- readRDS("~/results/FARES_plots/drug_info_sbae_drugs.rds")
drug_info_all <- readRDS("~/results/FARES_plots/drug_info_all.rds")
sbae_targets_df_top <- readRDS("~/results/FARES_plots/sbae_targets_df_top.rds")
```



## Analysis

```{r}
table(sbae_targets_df_top$p_adj < 0.001)
```


```{r}
SBAE_drug_targets <- unfactor(sbae_targets_df_top$sbae_targets)
```

remove missing names
```{r}
all_targets <- unlist(str_split(drug_info_all[,2], ";"))
all_targets <- all_targets[!is.na(all_targets )]
all_targets <- all_targets[!all_targets == "NA"]
```

```{r}
all_targets_df <- as.data.frame(table(all_targets))
```

```{r}
all_drug_targets <- unfactor(all_targets_df$all_targets)
```


core gene analysis 

```{r}
files <- list.files(path = "~/results/alpaca/alpaca_core_gene_lists")
```


```{r}
tissues<- str_split(files, "_female", simplify = TRUE)[,1]
tissues<- str_split(tissues, "_male", simplify = TRUE)[,1]
tissues<- unique(tissues)
```


Run permutation testing for male core genes
```{r}
pvalue <- c()
plots <- list()
core_gene_count <- c()
core_gene_list <- list()
for( i in 1:length(tissues)){
  print(tissues[i])
  male_core_genes <- readRDS(paste("~/results/alpaca/alpaca_core_gene_lists/", tissues[i],"_male_core_genes.rds", sep=""))
  
  sbae_res <- table(SBAE_drug_targets %in% unique(male_core_genes$symbol))
  
  sbae_count <- sbae_res[names(sbae_res) == "TRUE"]
  
  outcome <- drug_gene_core_test(sbae_count, all_drug_targets, male_core_genes$symbol)
  
  pvalue[i] <- outcome[1]
  plots[i] <- outcome[2]
  core_gene_count[i] <- sbae_count
  core_gene_list[[i]] <- SBAE_drug_targets[SBAE_drug_targets %in% unique(male_core_genes$symbol)]
}
```


```{r}
p_adj <-  p.adjust(pvalue, method = "BH")
```

```{r}
male_sex_core_gene <- data.frame(tissue = tissues, count_core_gene = core_gene_count)
```

```{r}
male_sex_core_gene <- cbind(male_sex_core_gene, unlist(pvalue), unlist(p_adj))
```

```{r}
male_sex_core_gene$fraction <- male_sex_core_gene$count_core_gene/84
```

```{r}
male_sex_core_gene_df_top <- male_sex_core_gene[order(male_sex_core_gene$count_core_gene),]
male_sex_core_gene_df_top$tissue <- factor(male_sex_core_gene_df_top$tissue, levels = male_sex_core_gene_df_top$tissue)
```

```{r}
male_sex_core_gene_df_top$sig <- ifelse(male_sex_core_gene_df_top$`unlist(p_adj)` < 0.05, "Yes", "No")
```

```{r}
male_sex_core_gene_df_top$tissue2 <- str_replace_all(male_sex_core_gene_df_top$tissue, "_", " ")
```

```{r}
male_sex_core_gene_df_top$tissue2 <- factor(male_sex_core_gene_df_top$tissue2, levels = male_sex_core_gene_df_top$tissue2)
```

```{r}
saveRDS(male_sex_core_gene_df_top, "~/results/alpaca/drug_target/male_sex_core_gene_df_top.rds")
```

```{r}
ggplot(male_sex_core_gene_df_top, aes(y=tissue2, x= fraction , label= count_core_gene, fill=sig))+ 
  geom_bar(stat="identity",color= "black") + 
  geom_text(size = 5, position = position_stack(vjust = 0.8)) +
  ylab("Tissue") + xlab("The fraction of SBAE drug targets\nthat are core genes") +
  theme(text = element_text(size = 20,  face="bold")) +
  scale_fill_viridis_d(option = "E", alpha= 0.7) + labs(fill="Significant?")

ggsave("~/results/alpaca/drug_target/230313_male_drug_target_core_gene_barplot.png", height=15, width = 10)
```

```{r}
male_drug_target_core_count <- as.data.frame(table(unlist(core_gene_list)))
```

make this plot

```{r}
male_drug_target_core_count <- male_drug_target_core_count[order(male_drug_target_core_count$Freq),]
male_drug_target_core_count$Var1 <- factor(male_drug_target_core_count$Var1, levels = male_drug_target_core_count$Var1)
```

```{r}
ggplot(male_drug_target_core_count, aes(y=Var1, x= Freq , label= Freq))+
  geom_bar(stat="identity",color= "black", fill = "white") +
  geom_text(size = 5, position = position_stack(vjust = 0.8)) +
  ylab("SBAE Drug Target") + xlab("The number of tissues with drug target as core gene") +
  theme(text = element_text(size = 20,  face="bold")) 
ggsave("~/results/alpaca/drug_target/230313_male_drug_target_core_gene_barplot_just_genes.png", height=15, width = 10)
```


Run permutation testing for female core genes
```{r}
pvalue<- c()
plots<- list()
core_gene_count <- c()
core_gene_list <- list()
for( i in 1:length(tissues)){
  print(tissues[i])
  female_core_genes <- readRDS(paste("~/results/alpaca/alpaca_core_gene_lists/", tissues[i], "_female_core_genes.rds", sep=""))
  
  sbae_res <- table(SBAE_drug_targets %in% unique(female_core_genes$symbol))
  
  sbae_count <- sbae_res[names(sbae_res) == "TRUE"]
  
  outcome <- drug_gene_core_test(sbae_count, all_drug_targets, female_core_genes$symbol)
  
  pvalue[i] <- outcome[1]
  plots[i] <- outcome[2]
  core_gene_count[i] <- sbae_count
  core_gene_list[[i]] <- SBAE_drug_targets[SBAE_drug_targets %in% unique(female_core_genes$symbol)]
}
```


```{r}
p_adj <- p.adjust(pvalue, method= "BH")
```

```{r}
female_sex_core_gene <- data.frame(tissue = tissues, count_core_gene = core_gene_count) 
```

```{r}
female_sex_core_gene <- cbind(female_sex_core_gene, unlist(pvalue), unlist(p_adj))
```

```{r}
female_sex_core_gene$fraction<- female_sex_core_gene$count_core_gene/84
```

```{r}
female_sex_core_gene_df_top <- female_sex_core_gene[order(female_sex_core_gene$count_core_gene), ]
female_sex_core_gene_df_top$tissue <- factor(female_sex_core_gene_df_top$tissue, levels = female_sex_core_gene_df_top$tissue)
```

```{r}
female_sex_core_gene_df_top$sig <- ifelse(female_sex_core_gene_df_top$`unlist(p_adj)` < 0.05, "Yes", "No")
```

```{r}
female_sex_core_gene_df_top$tissue2 <- str_replace_all(female_sex_core_gene_df_top$tissue, "_", " ")
```

```{r}

female_sex_core_gene_df_top$tissue2 <- factor(female_sex_core_gene_df_top$tissue2, levels = female_sex_core_gene_df_top$tissue2)
```

```{r}
saveRDS(female_sex_core_gene_df_top, "~/results/alpaca/drug_target/female_sex_core_gene_df_top.rds")
```


```{r}
ggplot(female_sex_core_gene_df_top, aes(y = tissue2, x = fraction , label = count_core_gene, fill = sig))+
  geom_bar(stat = "identity",color = "black") + 
  geom_text(size = 5, position = position_stack(vjust = 0.8)) + ylab("Tissue") +
  xlab("The fraction of SBAE drug targets\nthat are core genes") +
  theme(text = element_text(size = 20,  face = "bold")) +
  scale_fill_viridis_d(option = "E", alpha = 0.7) + labs(fill = "Significant?")
ggsave("~/results/alpaca/drug_target/230313_female_drug_target_core_gene_barplot.png", height=15, width = 10)
```

```{r}
female_drug_target_core_count <- as.data.frame(table(unlist(core_gene_list)))
```

make this plot

```{r}
female_drug_target_core_count <- female_drug_target_core_count[order(female_drug_target_core_count$Freq), ]
female_drug_target_core_count$Var1 <- factor(female_drug_target_core_count$Var1, levels = female_drug_target_core_count$Var1)
```

```{r}
ggplot(female_drug_target_core_count, aes(y = Var1, x = Freq , label = Freq))+
  geom_bar(stat = "identity",color = "black", fill = "white") + 
  geom_text(size = 5, position = position_stack(vjust = 0.8)) +
  ylab("SBAE Drug Target") + xlab("The number of tissues with drug target as core gene") +
  theme(text = element_text(size = 20,  face="bold")) 
ggsave("~/results/alpaca/drug_target/230313_female_drug_target_core_gene_barplot_just_genes.png", height=15, width = 10)
```



Now look at sex-bias expression 
gene list from the following GTEX study: https://www.science.org/doi/10.1126/science.aba3066#:~:text=The%20role%20of%20sex%20in%20the%20human%20transcriptome,-In%20humans%2C%20the&text=examined%20Genotype%2DTissue%20Expression%20(GTEx,in%20cellular%20composition%20across%20tissues.
```{r}
sex_bias_genes_gtex_science_paper <- read.table("~/data/GTEx_Analysis_v8_sbgenes/signif.sbgenes.txt", header=TRUE)
```

```{r}
table(sex_bias_genes_gtex_science_paper$tissue)
```

```{r}
#for each tissue find the male and female biased genes
tissues <- unique(sex_bias_genes_gtex_science_paper$tissue)
male_tissues_names <- paste("M", tissues, sep="_")
female_tissues_names <- paste("F", tissues, sep="_")
for (i in 1:length(tissues)){
  assign(male_tissues_names[i], sex_bias_genes_gtex_science_paper[sex_bias_genes_gtex_science_paper$tissue ==tissues[i] & sex_bias_genes_gtex_science_paper$effsize < 0, ])
  assign(female_tissues_names[i], sex_bias_genes_gtex_science_paper[sex_bias_genes_gtex_science_paper$tissue ==tissues[i] & sex_bias_genes_gtex_science_paper$effsize > 0, ])
}
```

```{r}
both_sex_tissue_names <- c(male_tissues_names, female_tissues_names)

geneset <- data.frame(id = both_sex_tissue_names, name = both_sex_tissue_names, genes = I(rep(list(c("a","b")), 88)))
sex_genes <- both_sex_tissue_names
for (i in 1:length(both_sex_tissue_names)){
  data <- get(both_sex_tissue_names[i])
  trim_names <- strtrim(data[,1],15)
  assign(sex_genes[i], trim_names)
  
}
sex_gene_set <- list(M_Adipose_Subcutaneous = M_Adipose_Subcutaneous, M_Adipose_Visceral_Omentum = M_Adipose_Visceral_Omentum)
```

```{r}
for (j in 1:88){
  sex_gene_set[[j]] <- get(sex_genes[j])
}
names(sex_gene_set) <-sex_genes
```


need to convert symbols to ensemble ids 

```{r}
SBAE_drug_targets_v2 <- gprofiler2::gconvert(SBAE_drug_targets, mthreshold = 1)
```

```{r}
all_drug_targets_v2 <- gprofiler2::gconvert(all_drug_targets, mthreshold = 1)
```

Run permutation testing for sex-bias expression
```{r}
pvalue <- c()
plots <- list()
expression_count <- c()
expression_gene_list <- list()

for( i in 1:length(sex_gene_set)){
  print(names(sex_gene_set)[i])
  
  expression_list <- sex_gene_set[[i]]

  sbae_res <- table(SBAE_drug_targets_v2$target %in% expression_list)
  
  sbae_count <- sbae_res[names(sbae_res) == "TRUE"]
  print(sbae_count)
  sbae_count <- ifelse(length(sbae_count)== 0, 0, sbae_count)
  outcome <- drug_gene_expression_test(sbae_count, all_drug_targets_v2$target, expression_list)
  
  pvalue[i] <- outcome[1]
  plots[i] <- outcome[2]
  expression_count[i] <- sbae_count
  expression_gene_list[[i]] <- SBAE_drug_targets_v2$target[SBAE_drug_targets_v2$target %in% expression_list]
}
```

```{r}
p_adj<-  p.adjust(pvalue, method= "BH")
```

```{r}
sex_expression_gene <- data.frame(tissue = names(sex_gene_set), expression_count = expression_count) 
```

```{r}
sex_expression_gene <- cbind(sex_expression_gene, unlist(pvalue), unlist(p_adj))
```

```{r}
sex_expression_gene$fraction <- sex_expression_gene$expression_count/84
```

```{r}
sex_expression_gene$sig <- ifelse(sex_expression_gene$`unlist(p_adj)` < 0.05, "Yes", "No")
```

```{r}
female_sex_expression_gene <- sex_expression_gene[grep("F_", sex_expression_gene$tissue), ]
```
```{r}
female_sex_expression_gene$tissue2 <- str_split(female_sex_expression_gene$tissue, "F_", simplify = TRUE)[,2]
female_sex_expression_gene$tissue2 <- str_replace_all(female_sex_expression_gene$tissue2, "_", " ")
```

```{r}
female_sex_expression_gene_df_top <- female_sex_expression_gene[order(female_sex_expression_gene$expression_count), ]
female_sex_expression_gene_df_top$tissue2 <- factor(female_sex_expression_gene$tissue2, levels = female_sex_expression_gene$tissue2)
```

```{r}
saveRDS(female_sex_expression_gene_df_top, "~/results/alpaca/drug_target/female_sex_expression_gene_df_top.rds")
```

```{r}
ggplot(female_sex_expression_gene_df_top, aes(y = tissue2, x = fraction , label = expression_count, fill = sig)) +
  geom_bar(stat = "identity",color = "black") + 
  geom_text(size = 5, position = position_stack(vjust = 0.8)) +
  ylab("Tissue") + xlab("The fraction of SBAE drug targets that have\nfemale-bias expression") +
  theme(text = element_text(size = 18,  face = "bold")) +
  scale_fill_viridis_d(option = "E", alpha = 0.7) + labs(fill = "Significant?")
ggsave("~/results/FARES_plots/230313_drug_target_female_bias_expression_barplot.png", height=15, width = 10)
```


```{r}
male_sex_expression_gene <- sex_expression_gene[grep("M_", sex_expression_gene$tissue), ]
```
```{r}
male_sex_expression_gene$tissue2 <- str_split(male_sex_expression_gene$tissue, "M_", simplify = TRUE)[,2]
male_sex_expression_gene$tissue2 <- str_replace_all(male_sex_expression_gene$tissue2, "_", " ")
```

```{r}
male_sex_expression_gene_df_top <- male_sex_expression_gene[order(male_sex_expression_gene$expression_count), ]
male_sex_expression_gene_df_top$tissue2 <- factor(male_sex_expression_gene$tissue2, levels = male_sex_expression_gene$tissue2)
```

```{r}
saveRDS(male_sex_expression_gene_df_top, "~/results/alpaca/drug_target/male_sex_expression_gene_df_top.rds")
```

```{r}
ggplot(male_sex_expression_gene_df_top, aes(y=tissue2, x= fraction , label= expression_count, fill=sig)) +
  geom_bar(stat="identity",color= "black") + 
  geom_text(size = 5, position = position_stack(vjust = 0.8)) +
  ylab("Tissue") + xlab("The fraction of SBAE drug targets that have\nmale-bias expression") +
  theme(text = element_text(size = 18,  face="bold")) + scale_fill_viridis_d(option = "E", alpha= 0.7) +
  labs(fill="Significant?")
ggsave("~/results/FARES_plots/230313_drug_target_male_bias_expression_barplot.png", height=15, width = 10)
```


```{r}
expression_drug_targets <- SBAE_drug_targets_v2[ SBAE_drug_targets_v2$target %in% unique(unlist(expression_gene_list)),]
```

```{r}
saveRDS(expression_drug_targets, "~/results/alpaca/drug_metabolism_genes/expression_drug_targets.rds")
```


# END
    Location of final scripts:
    src
    Location of data produced:
    ~/results/FARES_plots and ~/results/alpaca/drug_metabolism_genes/
    Dates when operations were done:
    230313
    
## Versions
```{r}
sessionInfo()
```

